Add `sub_searches` to the search endpoint

jdconrad opened the pull request:
This change adds a new top-level element to the search endpoint called `sub_searches`. This top-level element allows for a list of additional searches where each "sub search" will have a query executed separately as part of ranking and later combined into a final single set of documents based on the ranking algorithm.

The `sub_searches` element is used instead of `query` (and they may not be used together) to allow a search using ranking to execute multiple queries.

An example search using `sub_searches` looks like:

```
GET example-index/_search
{
    "sub_searches": 
        [
            {
                "query": {
                    "term": {
                        "text": "red shoes"
                    }
                }
            },
            {
                "query": {
                    "term": {
                        "text": "blue shoes"
                    }
                }    
            }
        ],
    "rank": {
        "rrf": {
            "window_size": 100,
            "rank_constant": 25
        }
    }
}
```

jdconrad committed:
add wrapper object for a single query

jdconrad committed:
add wrapper for search queries to carry enough information to do
execute multiple queries individually

jdconrad committed:
add paths for dfs and no dfs phases

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix npe

jdconrad committed:
another npe

jdconrad committed:
fix test

jdconrad committed:
fix npe

jdconrad committed:
make version high to avoid conflicts for now

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix transport version id

jdconrad committed:
add an it test for 2 bm25 queries as part of rrf

jdconrad committed:
fix test

jdconrad committed:
fix test

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
add restriction for query and queries used at the same time

jdconrad committed:
Add tests for SearchQueryBuilder

jdconrad committed:
rename SearchQuery to SearchQueryWrapper

jdconrad committed:
remove dead code

jdconrad committed:
do not use single shard shortcut when using rank

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
update shard counts

jdconrad committed:
add additional check for single shard optimization to not run with rank

jdconrad committed:
spotless

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
add logic for queries to can match

jdconrad committed:
spotless

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix can match issue with empty queries

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
add tests with multi bm25 queries and knn searches

jdconrad committed:
add rrf tests for coordinator can match

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
add tests for shard can match

jdconrad committed:
move cluster setup to single test

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
add some basic yaml test for rrf with multiple queries

jdconrad committed:
add it for linear combination with queries

jdconrad commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -135,7 +136,7 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
      * This should be the transport version with the highest id.
      */
-    public static final TransportVersion CURRENT = V_8_500_007;
+    public static final TransportVersion CURRENT = V_8_500_999;
```
I will update this once the PR is ready, but did not want to contend with other changes as this may need multiple rounds of review.

jdconrad committed:
resolve ambiguity in multi shard from identitical score

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @jdconrad, I've created a changelog YAML for you.

jdconrad committed:
Update docs/changelog/96224.yaml

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1485,6 +1536,19 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
                         }
                     }
                     searchUsage.trackSectionUsage(KNN_FIELD.getPreferredName());
+                } else if (QUERIES_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
```
It it ok to have only a single query object under `queries`?

jdconrad replied:
I think it makes sense to allow that if you wanted to sort on the individual query (future improvement), but not on a knn search as part of ranking.

mayya-sharipova replied:
Since we are modifying this file, should we also make sure that (de)/serialization  and from/toXcontent tests are covered in `SearchSourceBuilderTests` by adding `queries` part to `RandomSearchRequestGenerator`?

jdconrad replied:
Yes, I will add this there.

jdconrad replied:
I added this.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/builder/SearchQueryWrapperBuilder.java:
```diff
@@ -0,0 +1,115 @@
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (SearchSourceBuilder.QUERY_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
```
Should we fail add fail in other case (where QUERY_FIELD doesn't match), otherwise we accept arbitrary request like this:
```js
{
    "queries": [
        {
            "query": {
                "match": {
                    "k1": "blue"
                }
            }, 
            "rrf" : 2
        }
    ]
}
```

jdconrad replied:
Good catch! I'll update this to not accept extraneous parameters.

benwtrent replied:
I agree with Mayya, we should restrict this until we are sure more options are added.

jdconrad replied:
I've changed this to a ConstructingObjectParser which should create stronger consistency in both error messages and disallowing unknown fields.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -126,6 +130,22 @@ protected SearchPhase getNextPhase(final SearchPhaseResults<SearchPhaseResult> r
 
     private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
+        if (request.source() != null && request.source().queries().isEmpty() == false) {
+            BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
```
We probably not need to rewrite to `bool` if there is only 1 searchQueryWrapperBuilder, but also I assume this would be quite a rare case.

jdconrad replied:
Following up on the comment above, maybe we it does make sense to ensure that queries is only used with at least 2 of either queries or knn searches? So if there's only one `SearchQueryWrapper` it would also require that there's at least one knn search.

jdconrad replied:
This is no longer possible with the removal of default linear combination for multiple queries.

benwtrent commented on server/src/main/java/org/elasticsearch/search/builder/SearchQueryWrapperBuilder.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    public SearchQueryWrapperBuilder(StreamInput in) throws IOException {
+        this.queryBuilder = in.readOptionalNamedWriteable(QueryBuilder.class);
```
Why is this optional? This should always be non-null and we should ensure that. I don't think 

```
"queries": [{}]
```
Is valid.

jdconrad replied:
Good point. I will change this.

jdconrad replied:
I've changed to be a required field of a ConstructingObjectParser.

benwtrent reviewed (commented):
High level pass. I am gonna have to think about `queries` vs `query` :)

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -359,6 +359,9 @@ public ActionRequestValidationException validate() {
             }
         }
         if (source != null) {
+            if (source.query() != null && source.queries().isEmpty() == false) {
```
I have it set up to do linear combination by default since that's the same behavior for multiple knn queries now. Once things like sort and rescore are possibly added we would have to know we can't just shortcut them into a boolean query anymore.

benwtrent replied:
> Once things like sort and rescore are possibly added we would have to know we can't just shortcut them into a boolean query anymore.

If this is a future concern, we should disallow it now instead of switching behavior later.

jdconrad replied:
I apologize as I didn't explain all of my thoughts on this. I think what is in this PR now is the behavior that makes the most sense since we already default to linear combination for some combination of query and knn searches. What we can do in the future is see if queries is just a set of *only* QueryBuilders with no sorts, no re-scores, etc. and continue this default behavior, otherwise we can use an (interal?) linear combination ranker to execute each query separately like we do for RRF now. This should only be a bit of extra logic as part of `if/else` block.

jdconrad replied:
After more thought, I will go ahead and disallow this as both Ben and Luca suggest. I still think that it's more consistent to default to linear combination, but we can always add it later.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -353,10 +353,18 @@ protected final void innerWriteTo(StreamOutput out, boolean asKey) throws IOExce
+                    }
+                }
+                out.writeNamedWriteableList(rankQueryBuilders);
+            } else if (source != null && source.queries().isEmpty() == false) {
+                throw new IllegalArgumentException("cannot serialize [rank] to version [" + out.getTransportVersion() + "]");
+            }
```
Copy and paste error :)

jdconrad replied:
Changed from `rank` to `queries`.

jdconrad replied:
I just pushed. I was grouping multiple changes together for a single response.

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1485,6 +1536,19 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
+                        } else {
+                            throw new XContentParseException(
+                                parser.getTokenLocation(),
+                                "malformed queries format within the queries search array only objects are allowed; found " + token
```
I'll fix this.

jdconrad replied:
Updated.

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
response to pr comments

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java:
```diff
@@ -98,4 +102,19 @@ protected SearchPhase getNextPhase(final SearchPhaseResults<DfsSearchResult> res
         );
     }
+
+    private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
```
I actually just spoke about this same logic with @javanna earlier today. I admit this is not easy to follow, and I'm certainly open to suggestions for improvement. The logic is here to support accurately gathering terms statistics. We could instead go through each query potentially as part of the dfs phase on the shard. I'm not sure that's necessarily simpler. I think we would probably want to convert this to a BooleanQuery either way to try to avoid gathering multiple term stats on the shard. No matter what we decide, we do have to address this prior to having the knn queries ready to integrate.

javanna replied:
I'd try to have the linear combination in a single place, which is easier to follow and that is the query used for aggs, rather than requiring the linear combination like here for dfs, which does not include knn. I am not entirely sure about the implications, curious to hear your findings Jack.

jdconrad replied:
With the removal of default linear combination I was able to move this into a single method as part of `AbstractSearchAsyncAction`. The rewrite behavior is now identical for dfs-query-then-fetch and query-then-fetch.

jdconrad replied:
I apologize as I didn't explain this well. I meant for accurate term stats you still need to process multiple queries in `DFSPhase` in some way prior to processing completed knn searches in the `DFSQueryPhase`.

You are correct that we could iterate through multiple queries and combine the term stats for the same results. But that is more complex than having a single boolean query created on the coordinator prior to hitting the `DFSPhase` on the shards.

jdconrad replied:
I figured out a way to move this to the shard in a single location in `parseSource`. I described it in one my high-level comments below. Hopefully, this reduces some of the complexities here.

jdconrad committed:
fix validate test

jdconrad committed:
add queries parameter to random source builder for tests

jdconrad commented:
I've responded to all the feedback in this PR so far.

javanna commented on server/src/main/java/org/elasticsearch/search/builder/SearchQueryWrapperBuilder.java:
```diff
@@ -0,0 +1,115 @@
+ * as part of a series of multiple queries for features like ranking.
+ * It's expected to typically be used as part of a {@link java.util.List}.
+ */
+public class SearchQueryWrapperBuilder implements ToXContent, Writeable, Rewriteable<SearchQueryWrapperBuilder> {
```
do we want to consider including rank in their name (also in the name of the section) given they only make sense when a rank section is provided?

jdconrad replied:
I thought about adding `rank`, but I decided against that because it would be inconsistent with `queries` being a top-level element. Maybe something like `LimitedSearchSourceBuilder`, `ChildSearchSourceBuilder`, or `SubSearchSourceBuilder`?

javanna replied:
SubSearchSourceBuilder sounds good to me.

javanna replied:
Otherwise, wild idea: how about we only write a new parser that parses a SearchSourceBuilder?

javanna commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -278,10 +277,11 @@ public ShardSearchRequest(StreamInput in) throws IOException {
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0) && in.getTransportVersion().before(TransportVersion.V_8_500_004)) {
+            List<QueryBuilder> rankQueryBuilders = in.readNamedWriteableList(QueryBuilder.class);
+            for (QueryBuilder queryBuilder : rankQueryBuilders) {
+                source.queries().add(new SearchQueryWrapperBuilder(queryBuilder));
```
this is some confusing bit to me that I need to verify: we were previously serializing the separate queries as part of shard search request. they were the query + each knn query. Now that is supported upstream by SearchSourceBuilder as it supports multiple queries, but is knn also part of it and if so when does it become part of it? Do we have a bwc test that covers this scenario cross version?

jdconrad replied:
I will try to add more tests for this. If knn searches are specified, we know that we are doing a dfs-query-then-fetch type of search. This means that after the dfs phase returns, we rewrite the shard requests for knn searches to be queries. They are applied in the rewrite method of `DfsQueryPhase`. The knn searches are then removed from the shard requests which is how it was before as well.

jdconrad replied:
BWC for 8.8 covered this in multiple tests (a couple hundred failed :) ) and caught a bug here (the one that took a long time), so I would say coverage on this is good.

javanna reviewed (commented):
I left some initial comments, this looks good already, just a couple of cosmetic things.

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/SearchQueryWrapper.java:
```diff
@@ -0,0 +1,30 @@
+ * as part of a series of multiple queries for features like ranking.
+ * It's expected to typically be used as part of a {@link java.util.List}.
+ */
+public class SearchQueryWrapper {
```
Agreed! I removed this in favor of going back to Query.

jdconrad commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1623,14 +1625,17 @@ private static boolean canMatchAfterRewrite(final ShardSearchRequest request, fi
-        } else {
-            // null query means match_all
-            canMatch = aliasFilterCanMatch;
+            if (request.source().query() != null) {
```
I looked into this briefly, and I agree this would be a positive change, but I think it's complex enough that I would prefer to do it as a follow up PR.

jdconrad replied:
After looking into it a bit more, I think this is quite difficult in a subtle way where we already make certain assumptions if queries isn't empty. That's not to say it can't be done, just we replace query with something like queries is size of 1 and there's no knn searches, etc.

javanna replied:
This seems odd, I'd like to understand more around why we need that type of conditional. It's counter intuitive to me that if you have multiple queries, you need to do things substantially differently compared to when you have a single one.

jdconrad replied:
One other important note that I forgot to mention is we actually use `query` for an aggregation as part of rank. This is a really nice way to convey this information.

The most challenging part of this change is how to handle SearchContext and where I think you would end up doing checks to see if you only have 1 query or not. Maybe it is only one check, though for single query. You'll also need additional checks anywhere you process queries to see if it's 1 or not, but we already have checks anyway for if queries is empty, so maybe I overestimated how hard this would be.

If it's all right I would still prefer to try to make this change as a separate PR.

jdconrad replied:
This has been updated to only use `queries` now.

javanna replied:
thanks a lot for the effort you put into this change, it makes things a whole lot better!

jdconrad committed:
remove SearchQueryWrapper

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix tests

jdconrad committed:
require rank to use queries

jdconrad committed:
disallow linear combination with queries

jdconrad committed:
remove linear combination queries tests

jdconrad commented:
@javanna @benwtrent @mayya-sharipova Thank you all for the feedback so far! I've done my best to apply your suggestions and improve the code here. The biggest change so far is the removal of default linear combination when using `queries`. I believe this simplified several things throughout.

jdconrad committed:
Merge branch 'main' into rrfmulti

javanna commented on x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java:
```diff
@@ -0,0 +1,257 @@
+
+@ESIntegTestCase.ClusterScope(maxNumDataNodes = 3)
+@ESIntegTestCase.SuiteScopeTestCase
+public class RRFRankCoordinatorCanMatchIT extends ESIntegTestCase {
```
does this need multiple nodes or could it be a single node test case?

javanna replied:
I am thinking that single node test case should be the default, unless you need to start nodes or you specifically need multi node clusters. This makes things easier to reason about, and more reproducible. I don't think we would lose a lot in terms of coverage.

jdconrad replied:
I believe since we use between 1 to 3 nodes randomly, this covers both uses cases of a single shard - one where it's on the same node as the coordinator, and one where it's on a different node than the coordinator. Multi-shard does not cover single shard use cases now, so I think there is merit to keeping this as is.

javanna commented on x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorCanMatchIT.java:
```diff
@@ -0,0 +1,257 @@
+            .setQueries(
+                List.of(
+                    new SearchQueryWrapperBuilder(QueryBuilders.rangeQuery("@timestamp").gt(4000).lte(5000)),
+                    new SearchQueryWrapperBuilder(QueryBuilders.rangeQuery("@timestamp").gte(10000).lt(10005))
```
this is a good test! thanks for coming up with it. There is no chance that this can be a more contained test around can match? We have similar tests in `SearchServiceTests` I think

jdconrad replied:
This mimics portions of the can match tests for CCS. It's tricky to override the time based portion of collecting the specific `@timestamp` field, and this allows for that behavior fairly explicitly.

javanna commented on x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankMultiShardIT.java:
```diff
@@ -0,0 +1,971 @@
+
+@ESIntegTestCase.ClusterScope(maxNumDataNodes = 3)
+@ESIntegTestCase.SuiteScopeTestCase
+public class RRFRankMultiShardIT extends ESIntegTestCase {
```
what was wrong with the previous RRFRankIT ? Does this one require multiple nodes and the complexity that comes with it? I guess you were looking for end-to-end tests, would yaml tests also be ok?

javanna replied:
makes sense.

javanna commented on x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankShardCanMatchIT.java:
```diff
@@ -0,0 +1,251 @@
+
+@ESIntegTestCase.ClusterScope(maxNumDataNodes = 3)
+@ESIntegTestCase.SuiteScopeTestCase
+public class RRFRankShardCanMatchIT extends ESIntegTestCase {
```
Similar comment as above, can these tests be added to SearchServiceTests, or possibly to a new class together with the coordinator can match, that extends from single node test case?

javanna replied:
would we test a different codepath against a single node?

jdconrad replied:
Same comment as above. Given that we have shortcuts or may add shortcuts for different coordinator/data node relationships I would prefer to leave this as-is.

javanna commented on x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRankSingleShardIT.java:
```diff
@@ -0,0 +1,971 @@
+
+@ESIntegTestCase.ClusterScope(maxNumDataNodes = 3)
+@ESIntegTestCase.SuiteScopeTestCase
+public class RRFRankSingleShardIT extends ESIntegTestCase {
```
Is this test verifying the behaviour when multiple phases can be a single query and fetch phase? Can it be a single node test case? Can we have javadocs that describes what the test aims to verify?

javanna replied:
I cannot think of such a case, the main thing is that requests don't go through transport when executing everything on a single node, and that coordinator is always the same as data node, but we don't have conditionals around "do this if we have a single node".

jdconrad replied:
Changed to single node.

javanna reviewed (commented):
I did another round of review and left additional comments. Thanks for all the test coverage. A couple of additional scenarios to verify:

- are alias filters applied to all queries?
- is post_filter applied to all queries? 
- is min_score applied to all queries? 
- should we expand the search usage / query usage test to verify that the multiple queries are all tracked?

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -134,79 +134,61 @@ public void onFailure(Exception exception) {
+                source.queries(List.of());
             }
+            return request;
+        }
```
This has changed now a bit now that we disallow multiple queries from doing linear combination by default. There's still logic to handle linear combination with a query and knn searches, though.

jdconrad replied:
This is completely changed now by moving the boolean query creation to the shard.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -877,4 +879,22 @@ private synchronized Runnable tryQueue(Runnable runnable) {
+    protected static ShardSearchRequest rewriteShardSearchRequestQueries(ShardSearchRequest request) {
+        if (request.source() != null && request.source().queries().isEmpty() == false) {
+            assert request.source().query() == null && request.source().rankBuilder() != null;
+            assert request.source().queries().size() >= 2 || request.source().knnSearch().isEmpty() == false;
```
If you have a knn search, we may have only a single query as part of queries, but this would still be considered valid if you wanted to apply a rescore only to the query, but not the knn results.

javanna replied:
Let me clarify: the assert reads "either we have 2 or more queries, or at least one knn clause". I was wondering if knn is empty may we have 0, 1, or any number of query? Maybe a silly question, I was trying to figure out if the assert is extensive.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:
```diff
@@ -104,7 +104,9 @@ private void innerRun() throws Exception {
-        final boolean queryAndFetchOptimization = queryResults.length() == 1 && context.getRequest().hasKnnSearch() == false;
+        final boolean queryAndFetchOptimization = queryResults.length() == 1
+            && context.getRequest().hasKnnSearch() == false
+            && reducedQueryPhase.rankCoordinatorContext() == null;
```
This is specifically required for multiple queries. It wasn't missing before.

javanna replied:
I think what confuses me is that we don't look for the multiple queries but rather for the rank coordinator context. This could use a comment to explain why, or maybe the conditional could be shared in a single place, assuming it is always exactly the same.

jdconrad replied:
This is removed.

jdconrad replied:
I realized I didn't answer this correctly. This is required specifically for ranking. If we have a single query and multiple knn searches, those become a boolean query and it's fine to shortcut.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:
```diff
@@ -236,7 +236,8 @@ public void sendExecuteQuery(
         // this used to be the QUERY_AND_FETCH which doesn't exist anymore.
-        final boolean fetchDocuments = request.numberOfShards() == 1;
+        final boolean fetchDocuments = request.numberOfShards() == 1
+            && (request.source() == null || request.source().rankBuilder() == null);
```
Same as above. This is specifically required for multiple queries. It wasn't missing before.

jdconrad replied:
I realized I also didn't answer this correctly. New answer is in the above comment.

jdconrad commented on x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/100_rank_rrf.yml:
```diff
@@ -82,3 +87,94 @@ setup:
+        body:
+          track_total_hits: true
+          fields: [ "text" ]
+          queries: [
```
I just want to be sure that we don't anticipate this will ever be used for anything other than rank?

javanna replied:
true, it's hard to predict the future :)

jdconrad replied:
What about `sub_searches`?

benwtrent replied:
@jdconrad considering we are calling their Java representation `sub_searches`, I could be persuaded for that name ;)

jdconrad replied:
Renamed to `sub_searches`.

jdconrad commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -626,7 +627,7 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
                 tracer.stopTrace();
             }
-            if (request.numberOfShards() == 1) {
+            if (request.numberOfShards() == 1 && (request.source() == null || request.source().rankBuilder() == null)) {
```
This avoids short cutting to fetch if were ranking because we still need to do the final ranking on the coordinator. I considered doing the ranking here and allowing the shortcut, but when I tried it, it was quite a bit more complexity.

javanna replied:
Got it, that makes sense. Would it make sense to have a method somewhere called "canShortcutFetch" or a comment that explains the additional conditionals?

jdconrad replied:
I think both are good ideas, and I will tackle this in a follow up if that's okay.

jdconrad commented:
@javanna Answers for your questions:

    are alias filters applied to all queries?

Yes. This is the same behavior as query + knn searches. We could potentially override this as part of a sub-search, but this could be tricky with aggregations.

    is post_filter applied to all queries?

Same answer as above.

    is min_score applied to all queries?

Same answer as above.

    should we expand the search usage / query usage test to verify that the multiple queries are all tracked?

Yes. More tests! :)

jdconrad committed:
rename SearchQueryWrapperBuilder to SubSearchQueryBuilder

jdconrad committed:
spotless

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix transport version

jdconrad committed:
change name

jdconrad committed:
fix accidental line deletion

javanna commented:
Thanks for the clarifications @jdconrad , do we already have tests for alias filters, post_filter and min_score applied to multiple queries?

jdconrad committed:
fix unit tests

jdconrad committed:
move multi query logic to shards

jdconrad commented:
@javanna We do not have any of these tests specifically around RRF. It's now on my list to add some.

I also changed the code @benwtrent and you pointed out as complex to the shard. I moved the creation of the boolean query to where we parseSource so it's in one location always which I think covers both dfs and non-dfs queries. Now for knn we just add them to queries always when we rewrite on the coordinator. This is a good first step of always just using queries.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -138,75 +138,51 @@ ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
-                    }
+        source = source.shallowCopy();
+        source.query(boolQueryBuilder);
+        source.queries(searchQueryWrapperBuilders);
```
This moves everything to queries now in this location and for builds the appropriate query in `SearchService.parseSource` on the shard for the `SearchContext`.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -138,75 +138,51 @@ ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
-            }
-            request.source(newSource);
+        if (source.queries().isEmpty() == false) {
+            boolQueryBuilder = (BoolQueryBuilder) source.query();
```
This is changed now.

jdconrad committed:
fix bwc issues and update SearchSourceBuilder to only use queries as a
backing data structure

jdconrad commented:
@javanna I took another step toward trying to only use queries. In `SearchSourceBuilder` only queries exists as the backing data structure. I left the `query` getter and setter, though, as that is a much broader change that I think is left to a better PR. This also helped with simplifying transport for `ShardSearchRequest` and `SearchSourceBuilder`.

jdconrad committed:
fix named write

jdconrad committed:
fix validation test

jdconrad committed:
Merge branch 'main' into rrfmulti

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -129,7 +131,9 @@ public static HighlightBuilder highlight() {
     }
 
-    private QueryBuilder queryBuilder;
+    // private QueryBuilder queryBuilder;
```
should we remove this line?

jdconrad replied:
Yes! You're good at catching those :)

jdconrad committed:
remove extraneous commented out variable

jdconrad committed:
fix serialization test

jdconrad committed:
fix SearchSourceBuilder equals and hashcode

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-1

jdconrad committed:
only write rank query builders to prior verisons if rank is set

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix unit test

jdconrad committed:
spotless

benwtrent reviewed (commented):
Another comment. 

I really like the simplifications of pushing a single `query` into the subsearches collection. Helps clean up execution paths and reasoning.

mayya-sharipova reviewed (approved):
@jdconrad  Thanks Jack,  great work!!!! And I also liked how the code got simpler with this change!

One thing that bothers me is that we run queries two times (separately and as a part of boolean query). For expensive queries this could be significant. May be later (not as part of this PR) we can look on ways to optimize this.

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
fix serialization bug

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,46 @@ public void writeTo(StreamOutput out) throws IOException {
-    public SearchSourceBuilder query(QueryBuilder query) {
-        this.queryBuilder = query;
+    public SearchSourceBuilder query(QueryBuilder queryBuilder) {
+        subSearchSourceBuilders = new ArrayList<>();
+        if (queryBuilder != null) {
+            subSearchSourceBuilders.add(new SubSearchSourceBuilder(queryBuilder));
+        }
```
This follows the model we use for knn searches and removes the ambiguity between `null` vs `isEmpty`.

jdconrad committed:
rename queries to sub_searches

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,46 @@ public void writeTo(StreamOutput out) throws IOException {
 
     /**
-     * Gets the query for this request
+     * Gets the query for this request.
```
Added.

jdconrad committed:
add additional comment

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad commented:
Addressed the remaining comments, and pushing the following items as follow ups:
* additional tests to cover some gaps
* possibly switch more things to single node
* clean up fetch shortcut related to ranking

benwtrent commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -359,6 +359,9 @@ public ActionRequestValidationException validate() {
         }
         if (source != null) {
+            if (source.subSearches().size() >= 2 && source.rankBuilder() == null) {
+                validationException = addValidationError("[queries] requires [rank]", validationException);
```
```suggestion
                validationException = addValidationError("[sub_searches] requires [rank]", validationException);
```

jdconrad replied:
Fixed.

benwtrent commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -378,10 +381,10 @@ public ActionRequestValidationException validate() {
                     validationException = addValidationError(
-                        "[rank] requires a minimum of [2] result sets which"
-                            + " either needs at minimum [a query and a knn search] or [multiple knn searches]",
+                        "[rank] requires a minimum of [2] result sets using a combination of queries and/or knn searches",
```
now that its called `sub_searches` you need to update this error I think.

jdconrad replied:
Fixed.

benwtrent commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1485,6 +1553,23 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
+                        } else {
+                            throw new XContentParseException(
+                                parser.getTokenLocation(),
+                                "malformed query within the [queries] field; found " + token
```
Needs a rename

jdconrad replied:
Fixed.

benwtrent commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -353,10 +362,15 @@ protected final void innerWriteTo(StreamOutput out, boolean asKey) throws IOExce
+                }
+            }
             out.writeNamedWriteableList(rankQueryBuilders);
-        } else if (rankQueryBuilders.isEmpty() == false) {
-            throw new IllegalArgumentException("cannot serialize [rank] to version [" + out.getTransportVersion() + "]");
```
did you mean to delete this?

jdconrad replied:
Yes. This logic was moved to `SearchSourceBuilder` instead since where that's where `subSearches` lives.

jdconrad committed:
fix error messages

jdconrad committed:
replace more messages with sub searches in place of queries

javanna commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,52 @@ public void writeTo(StreamOutput out) throws IOException {
-        this.queryBuilder = query;
+    public SearchSourceBuilder query(QueryBuilder queryBuilder) {
+        subSearchSourceBuilders = new ArrayList<>();
+        if (queryBuilder != null) {
```
do we run into troubles if we enforce this to be non null? Was it used before to reset a previously set query perhaps? If so doesn't the new impl change that behaviour?

jdconrad replied:
This mimics the behavior prior by where an empty sub searches list is the same a null query.

javanna commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,52 @@ public void writeTo(StreamOutput out) throws IOException {
+            BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+            for (SubSearchSourceBuilder subSearchSourceBuilder : subSearchSourceBuilders) {
+                boolQueryBuilder.should(subSearchSourceBuilder.getQueryBuilder());
+            }
```
I am a bit doubtful about many existing callers of this method, that now end up retrieving the linear combination if sub_searches are provided. I can't decide if that's good (because they now get the boolean query automatically) or bad (because we have not explicitly decided that the linear combination applies to each of these cases). 

I wonder if this method should throw whenever it's called when there is more than one query, and we should introduce a new method that builds the linear combination so we track where that is specifically needed. Would that complicate things?

jdconrad replied:
I think as a follow up we should remove the query getter/setter, and force all callers to use subsearches to delegate to them what should be appropriate behavior, but for now this is what we use in QueryPhase to get the aggregations query for ranking or get a query for linear combination when hybrid search is used.

javanna replied:
I would even be ok leaving the single getter, but I think it's important to be explicit about when you need the linear combination vs when you expect to have single query: gut feeling at least.

javanna commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1485,6 +1553,23 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
+                            );
+                        }
+                    }
+                    searchUsage.trackSectionUsage(SUB_SEARCHES_FIELD.getPreferredName());
```
there's quite some API other than search that parse part of their content into a SearchSourceBuilder. Is it ok that they all end up supporting rank and sub_searches automatically?

jdconrad replied:
I think it should work for any API that wants to use the search endpoint, but it's possible I'm missing something.

javanna commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1623,14 +1626,12 @@ private static boolean canMatchAfterRewrite(final ShardSearchRequest request, fi
-            // null query means match_all
-            canMatch = aliasFilterCanMatch;
+            canMatch &= request.source()
+                .subSearches()
```
will subsearches here already contain the knn queries too? I think so, and that should not be an issue, but I wanted to double check.

jdconrad replied:
there should never be a knn search here because can_match only works with query_then_fetch, but not dfs_query_Then_fetch.

javanna replied:
oh good point. maybe that is something that we need to discuss changing.

jdconrad committed:
fix validation test

javanna reviewed (approved):
I left a bunch of new comments, mostly cosmetic things, or around hardening certain details. No blockers though, those could also be addressed as a follow-up especially if they require significant effort.

LGTM!

elasticsearchmachine commented:
Hi @jdconrad, I've updated the changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

jdconrad committed:
Update docs/changelog/96224.yaml

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,52 @@ public void writeTo(StreamOutput out) throws IOException {
-    public SearchSourceBuilder query(QueryBuilder query) {
-        this.queryBuilder = query;
+    public SearchSourceBuilder query(QueryBuilder queryBuilder) {
+        subSearchSourceBuilders = new ArrayList<>();
```
I admit this is a bit fragile, but I think this should be addressed as a follow up. The prior behavior would replace the previous query so asserting there would break updates to this method.

javanna replied:
I see, maybe just assert that there is a single element in the list then so it's 1:1 with the previous behaviour of replacing?

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -353,20 +368,52 @@ public void writeTo(StreamOutput out) throws IOException {
+     * Sets the queries for this request.
+     */
+    public SearchSourceBuilder subSearches(List<SubSearchSourceBuilder> subSearchSourceBuilders) {
+        this.subSearchSourceBuilders = subSearchSourceBuilders;
```
Yes! Good call.

jdconrad commented on server/src/test/java/org/elasticsearch/action/search/SearchRequestTests.java:
```diff
@@ -303,10 +329,9 @@ public void testValidate() throws IOException {
             assertEquals(
-                "[rank] requires a minimum of [2] result sets which "
-                    + "either needs at minimum [a query and a knn search] or [multiple knn searches]",
+                "[rank] requires a minimum of [2] result sets using a combination of queries and/or knn searches",
```
Fixed.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -138,75 +137,23 @@ ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
-
-            SearchSourceBuilder newSource = source.shallowCopy().query(searchQuery).knnSearch(List.of());
-            request.source(newSource);
-            request.rankQueryBuilders(rankQueryBuilders);
```
All of this logic goes away since we build the boolean query on the shard when necessary as part of parsing the `SearchSourceBuilder` into a `DefaultSearchContext`. I admit I was surprised by how much code we were able to pull from the coordinator phases. Best part is it naturally works for dfs, query, and fetch :)

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -103,12 +103,12 @@ static void executeRank(SearchContext searchContext) throws QueryPhaseExecutionE
                 break;
             }
-            RankSearchContext rankSearchContext = new RankSearchContext(searchContext, rankQuery, rankShardContext.windowSize());
+            RankSearchContext rankSearchContext = new RankSearchContext(searchContext, query, rankShardContext.windowSize());
```
Reverted. I had changed this when we had the wrapper object, but that was removed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -103,12 +103,12 @@ static void executeRank(SearchContext searchContext) throws QueryPhaseExecutionE
         int nodeQueueSize = querySearchResult.nodeQueueSize();
 
         // run each of the rrf queries
```
Fixed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -58,9 +58,9 @@ public class RankSearchContext extends SearchContext {
+    public RankSearchContext(SearchContext parent, Query query, int windowSize) {
         this.parent = parent;
-        this.rankQuery = parent.buildFilteredQuery(rankQuery);
+        this.rankQuery = parent.buildFilteredQuery(query);
```
Reverted. I had changed this when we had the wrapper object, but that was removed.

jdconrad committed:
response to pr comments

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -278,10 +277,20 @@ public ShardSearchRequest(StreamInput in) throws IOException {
-            rankQueryBuilders = List.of();
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0) && in.getTransportVersion().before(TransportVersion.V_8_500_999)) {
+            List<QueryBuilder> rankQueryBuilders = in.readNamedWriteableList(QueryBuilder.class);
+            if (rankQueryBuilders.size() == 1) {
```
I admit the checks here are odd. We have to account for the dfs phase in 8.8 where we can have a single query, and empty rankQueryBuilders even if we are ranking. We can never have a exactly one rank query builder, though, so this ensures that if that situation happens we will be able to effectively debug why through a stack trace.

jdconrad replied:
I changed this to an assert instead.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -278,10 +277,20 @@ public ShardSearchRequest(StreamInput in) throws IOException {
+            if (rankQueryBuilders.size() == 1) {
+                throw new IllegalStateException("[rank] requires at least [2] queries, but only found [1]");
+            }
+            if (source != null && source.rankBuilder() != null) {
```
Good point! I have changed this check to be `(rankQueryBuilders >= 2)` with an assert instead.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -278,10 +277,20 @@ public ShardSearchRequest(StreamInput in) throws IOException {
+                for (QueryBuilder queryBuilder : rankQueryBuilders) {
+                    subSearchSourceBuilders.add(new SubSearchSourceBuilder(queryBuilder));
+                }
+                if (subSearchSourceBuilders.size() >= 2) {
```
I changed this.

jdconrad committed:
cleaned up bwc deserialization

jdconrad committed:
fix change log with highlight

jdconrad committed:
Merge branch 'main' into rrfmulti

jdconrad committed:
Update 96224.yaml

Missing word in changelog

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-1

jdconrad commented:
@elasticmachine run elasticsearch-ci/part-1

jdconrad committed:
no op

jdconrad committed:
mute bad test

jdconrad commented:
Thank you again @mayya-sharipova, @benwtrent, and @javanna for the reviews!
